adding how to minimize CVE risk doc + zero-day glossary entry (#1143)
chainguard-dev/edu: content/chainguard/chainguard-images/cve-risk.md

---
title: "Strategies for Minimizing your CVE Risk"
linktitle: "Minimize CVE Risk"
type: "article"
description: "A conceptual article outlining best practicies for reducing one's CVE risk."
date: 2023-11-16T11:07:52+02:00
lastmod: 2023-11-16T11:07:52+02:00
draft: false
tags: ["Conceptual", "Chainguard Images", "CVE"]
images: []
menu:
  docs:
    parent: "chainguard-images"
weight: 740
toc: true
---

[Common vulnerabilities and exposures](/software-security/glossary/#cve), often referred to as "CVEs", are an increasingly frequent concern for both developers and consumers of software. A new CVE that appears in a widely-used application or a vulnerability scan with a large number of positive results would naturally be worrisome for developers, CISOs, and customers alike. 

> There's no way to guarantee that your software will never be impacted by a CVE. After all, there are software vulnerabilities currently in existence that haven't even been discovered (known as [zero-day vulnerabilities](/software-security/glossary/#zero-day)). With that said, this conceptual article aims to highlight a few practices and strategies you and your team can use to reduce the risk of CVEs on your software.

Reviewer: Can you add a sentence here (it can be a new paragraph if you want) that points to the `[recommended tools](#recommended-tools)` section?

## Maintaining dependency hygiene

Dependency hygiene is the practice of ensuring that the dependencies you introduce into your project do not contain vulnerabilities. We've found that many popular container images, when not updated, [will accumulate about one known vulnerability per day](https://www.chainguard.dev/unchained/enforce-against-vulnerability-sprawl-with-up-to-date-images). Many of the most widely used container images aren't updated regularly and past academic software engineering research has found that most images on Docker Hub, the most widely used container registry, have not been updated for over 120 days.

Some teams go so far as to actively avoid using dependencies, working under the belief that taking a "roll-your-own" approach to all their tooling is an effective way to reduce CVEs. Although first-party tools may report fewer CVEs when scanned, they won't be battle-tested like widely used open-source software, meaning there are drawbacks to this approach.

> ## Recommended tools

Reviewer: In my opinion, this section is the meat of the tutorial. 

To make this easier to parse, I think it would benefit from some H3 headings (scanners, automated updates, metrics reporting, containers). You can also bold the tools you're recommending.

Under the headings, you can add more general guidance on scanners, automated updates, scorecards, and containers. With scanners, I would include additional scanners also, and under the containers you can talk about Wolfi in addition to Images.

> There are a number of tools available that allow you to scan your third party code for CVEs. At Chainguard, we use [Grype](https://github.com/anchore/grype) to scan our own Chainguard Images, as it's open-source and it can scan Software Bills of Materials (or SBOMs). Additionally, Grype biases towards false positives over false negatives. This suits our needs, as we'd rather look into a vulnerability in one of our images that turns out to be a false positive than overlook a real vulnerability that impacts our customers.

Reviewer: I think you can also link out to the blog on this: https://www.chainguard.dev/unchained/why-chainguard-uses-grype-as-its-first-line-of-defense-for-cves

One tool that's useful for automating updates for your dependencies is [Dependabot](https://docs.github.com/en/code-security/dependabot). Dependabot is a free tool offered by GitHub that sends alerts to repositories affected by new vulnerabilities and — if possible — raises a pull request to update the affected dependencies.

It's important to treat every new dependency you introduce into an application with a degree of scrutiny. Although the needs of every project vary, there are a number of practices a team can implement to ensure that they only use dependencies that meet certain requirements. As an example, you could integrate a tool like [Scorecard](https://securityscorecards.dev/) into your organization's workflow.

Scorecard is an automated tool from the Open Source Security Foundation that performs a variety of security checks on software, returning a score between 1 and 10 for each one. These scores can help you understand what you need to work on to improve your project's security, and can also help you assess the security of your dependencies. 

> [Chainguard Images](https://www.chainguard.dev/chainguard-images) are also helpful for minimizing your CVE risk. Many of Chainguard's images are distroless, meaning they contain only an application and its runtime dependencies. These images do not even contain a shell or package manager. Because Chainguard Images minimize the number of dependencies in this manner, they inherently contain few to zero CVEs. Also, Chainguard Images are rebuilt nightly to ensure images are completely up-to-date and contain all available security patches.

Reviewer: ```suggestion
[Chainguard Images](https://www.chainguard.dev/chainguard-images) can minimize your CVE risk. Many of Chainguard's images are distroless, meaning they contain only an application and its runtime dependencies. These images do not even contain a shell or package manager. Because Chainguard Images minimize the number of dependencies in this manner and thus reducing the potential attack surface, they inherently contain few to zero CVEs. Also, Chainguard Images are rebuilt nightly to ensure images are completely up-to-date and contain all available security patches, and the engineering team [often fixes vulnerabilities before they're detected](https://www.chainguard.dev/unchained/how-chainguard-fixes-vulnerabilities).
```

## Learn more

> As mentioned in the introduction, there's no way to guarantee that no CVEs will affect your software. However, we hope that by reading this article you'll have learned some practical tips you can use to help minimize your exposure to vulnerabilities and keep your software secure. 

Reviewer: ```suggestion
As mentioned in the introduction, there's no way to guarantee that no CVEs will affect your software. However, we hope that by reading this article you'll have learned some practical tips you can use to help minimize your exposure to vulnerabilities and keep your software secure. 

```

Some teams go so far as to actively avoid using dependencies, working under the belief that taking a "roll-your-own" approach to all their tooling is an effective way to reduce CVEs. Although first-party tools may report fewer CVEs when scanned, they won't be battle-tested like widely used open-source software, meaning there are drawbacks to this approach.

## Recommended tools

This section outlines a few categories of tools that can be useful for minimizing CVEs, and provides a few examples for each.

### Scanners

There are a number of tools available that allow you to scan your third party code for CVEs. 

At Chainguard, we use [**Grype**](https://www.chainguard.dev/unchained/why-chainguard-uses-grype-as-its-first-line-of-defense-for-cves) to scan our own Chainguard Images, as it's open-source and it can scan Software Bills of Materials (or SBOMs). Additionally, Grype biases towards [false positives](/chainguard/chainguard-images/scanners/false-results/) over false negatives. Looking into a vulnerability in an image that turns out to be a false positive can be preferable to overlooking a real vulnerability that impacts end users.

> Another open-source scanning option is [**Falco**](https://falco.org/). Falco works with environments running in individual containers, hosts, Kubernetes, and the cloud. Falco works by collecting data from various sources — including Linux kernal syscalls, Kubernetes audit logs, and events from systems like GitHub or Okta — and compares them with a set of rules. Falco comes with a list of rules by default, but you can also create your own rules to suit the needs of yoru project. If any of the collected data breaks one of the rules, Falco will identify it as a security issue. 

Author: that does it, I'm finding a spellchecker for my IDE

There are a number of important considerations one should make when keeping container images up to date. Please check out our [conceptual article on the subject](/chainguard/chainguard-images/considerations-for-image-updates/).

### Minimal container images

As mentioned previously, the fewer dependencies a given piece of software uses, the lower likelihood that it will be impacted by CVEs. To this end, Chainguard provides a library of minimal container images that can minimize your CVE risk. 

[**Chainguard Images**](https://www.chainguard.dev/chainguard-images) are distroless, meaning they contain only an application and its runtime dependencies. These images do not even contain a shell or package manager. Because Chainguard Images minimize the number of dependencies in this manner and thus reduce the potential attack surface, they inherently contain few to zero CVEs. Also, Chainguard Images are rebuilt nightly to ensure images are completely up-to-date and contain all available security patches, and the engineering team [often fixes vulnerabilities before they're detected](https://www.chainguard.dev/unchained/how-chainguard-fixes-vulnerabilities).

> If you're looking for a certain image that isn't included Chainguard's library of Images, you can build your own minimal container images with [**Wolfi**](https://github.com/wolfi-dev), a community Linux distribution designed by Chainguard for the container and cloud-native era. Chainguard started the Wolfi project to enable building Chainguard Images, which required a Linux distribution with components at the appropriate granularity and with support for [glibc](https://www.gnu.org/software/libc/).

Reviewer: ```suggestion
If you're looking for a certain image that isn't included in Chainguard's library, you can build your own minimal container images with [**Wolfi**](https://github.com/wolfi-dev), a community Linux distribution designed by Chainguard for the container and cloud-native era. Chainguard started the Wolfi project to enable building Chainguard Images, which required a Linux distribution with components at the appropriate granularity and with support for [glibc](https://www.gnu.org/software/libc/).
```
